src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1170,7 +1193,9 @@ TaskListRequires2PC(List *taskList)
 static bool
 ReadOnlyTask(TaskType taskType)
 {
-	if (taskType == ROUTER_TASK || taskType == SQL_TASK)
+	if (taskType == ROUTER_TASK || taskType == SQL_TASK || taskType ==
+		MAP_OUTPUT_FETCH_TASK ||
+		taskType == MAP_TASK || taskType == MERGE_TASK)
```

Comment:
Maybe make it a case statement.

Minor: I could see us end up wanting to re-partition the output of a DELETE .. RETURNING at some point. I wonder whether we should start keeping an explicit readOnly flag.